### DISCUSSION TITLE
Source Code Debugging for Linux based apps on the MSS

### DISCUSSION THREAD

**Discussion by rockoguido** [polarfire-soc/polarfire-soc-documentation#98] [Linux build systems for PolarFire SoC] [answered]
Hi.  working with a development team that is using the PolarFire SOC.   Part of the team wishes to develop applications for the MSS and will use Microchip's Linux distro as the OS of choice on the MSS.  Using the Icicle Kit as the target platform.  Linux distro in the near term will be the image that comes pre-loaded on the Icicle Kit (Yocto)
At this time I have delivered a solution using SoftConsole to build the applications (using riscv64-unknown-linux-gnu-gcc) and application can be built and downloaded to the target via scp and executed without issue.    Have tested with a few networking based apps.  

riscv64-unknown-linux-gnu-gcc comes from Buildroot, so there is a mismatch between tools and target at the moment.  

Goal is gdbserver based debugging and to get away from current printf solution.  

Don't believe SoftConsole has the means to use GDB and have to transition to Eclipse, just want to verify that for starters.   

In addition I did build a Yocto distro for the Icicle Kit with the -c populate_sdk switch and have installed the cross tools (riscv64-oe-linux-)  
Built my app with these tools in Eclipse but it was much clunkier as opposed to using the buildroot tools.   i.e. manually setting include and lib paths etc. 

Without diving too deep just wanted to see if anyone has been down this path.   My inclination right now is to take the easier path that gets me to true source code debugging and align  target image with what cross tools work best.   

thx for listening

**Comment by AntonKrug** [comment, marked as answer]
> Don't believe SoftConsole has the means to use GDB and have to transition to Eclipse, just want to verify that for starters.

SoftConsole supports only the bare-metal targets. However, nothing is stopping you from making a separate installation of SoftConsole and changing the settings to use a Linux toolchain/gdb at your own risk.

**Reply by rockoguido** [reply]
yes it seems reading up on OpenOCD that there is a way to connect to a GDBSERVER session on the target.   I have an initial connection but debugging still an issue, will keep digging.

**Reply by AntonKrug** [reply]
If you would have a gdbserver running on a remote device over the network (like your icicle target), then you do not need OpenOCD (and its launchers). You can go in the SoftConsole into the debug configurations, click on the 'funnel' icon which opens a filter dialog where you can un-check the C/C++ Remote Application and then it will allow you to make the remote application launchers.

![filter](https://user-images.githubusercontent.com/30338596/175573361-2c282a06-164a-411a-8f72-ec874450476b.png)

**Reply by rockoguido** [reply]
heck yeah!!  that did it!  thanks for the great tip.    at this point I have a debug config that bypasses OPENOCD and just does straight TCP/IP to the target, nothing fancy, you still need to have the application and gdbserver on the target but this gives you source code debugging, a great improvement.   I'll put together an app note, 
thanks again!

**Reply by lhanlyu** [reply]
Also the yocto image 'mpfs-dev-cli' has gdb pre-loaded which you can use.

**Reply by AntonKrug** [reply]
Have a look at options the launchers provide. I think there was a way to tell the launcher that your sessions is remote and allow the launcher connect thorugh SSH to the target and do the gdbserver stuff for you. But at that point you might be actually better of using Eclipse CDT, as you see SoftConsole has selectively hidden/disabled some stuff to be focused on bare metal.

**Reply by rockoguido** [reply]
at this point I have a remote connection, just using TCP, bypassing OpenOCD, using SoftConsole and I can do a traditional source code debugging session on the target, this is dependent on having gdbserver running on the target.    My only hangup at the moment is that I'm using cross tools that are from Buildroot, had some issues using Yocto based tools that I'll sort out when I get a chance.   Goal is to document this in some type of App Note when done.
